### PR TITLE
correctly handle const params in type_of

### LINKED ISSUE #70273: ICE when using const generics with associated items
### Code

```rust
#![feature(const_generics)]

trait T<const A: usize> {
    type Item;
}
struct S;

impl<const A: usize> T<A> for S {
    type Item = i32;
}

fn main() {
    let _err: <S as T<0usize>>::Item;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.44.0-nightly (f509b26a7 2020-03-18)
binary: rustc
commit-hash: f509b26a7730d721ef87423a72b3fdf8724b4afa
commit-date: 2020-03-18
host: x86_64-unknown-linux-gnu
release: 1.44.0-nightly
LLVM version: 9.0

```

### Error output

```
   Compiling bug v0.1.0 (/tmp/bug)
warning: the feature `const_generics` is incomplete and may cause the compiler to crash
 --> src/main.rs:1:12
  |
1 | #![feature(const_generics)]
  |            ^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default

warning: Error finalizing incremental compilation session directory `/tmp/bug/target/debug/incremental/bug-19qm6gngth5a1/s-flslgmhmjy-18wc5w7-working`: No such file or directory (os error 2)

error: internal compiler error: mir_const_qualif: MIR had errors
  --> src/main.rs:13:23
   |
13 |     let _err: <S as T<0usize>>::Item;
   |                       ^^^^^^

error: internal compiler error: PromoteTemps: MIR had errors
  --> src/main.rs:13:23
   |
13 |     let _err: <S as T<0usize>>::Item;
   |                       ^^^^^^

error: internal compiler error: broken MIR in DefId(0:12 ~ bug[538c]::main[0]::{{constant}}[0]) ("return type"): bad type [type error]
  --> src/main.rs:13:23
   |
13 |     let _err: <S as T<0usize>>::Item;
   |                       ^^^^^^

error: internal compiler error: broken MIR in DefId(0:12 ~ bug[538c]::main[0]::{{constant}}[0]) (LocalDecl { mutability: Mut, local_info: Other, internal: false, is_block_tail: None, ty: [type error], user_ty: UserTypeProjections { contents: [] }, source_info: SourceInfo { span: src/main.rs:13:23: 13:29, scope: scope[0] } }): bad type [type error]
  --> src/main.rs:13:23
   |
13 |     let _err: <S as T<0usize>>::Item;
   |                       ^^^^^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:360:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.44.0-nightly (f509b26a7 2020-03-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: could not compile `bug`.

To learn more, run the command again with --verbose.


```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:360:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1427
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:515
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::drop_in_place
  15: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
  16: core::ptr::drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
```

</p>
</details>

This issue has been assigned to @lcnr via [this comment](https://github.com/rust-lang/rust/issues/70273#issuecomment-602208970).

### PR DESCRIPTION
**Pull request by lcnr**
extends #70223, retry of #70276

fixes #70273

r? @eddyb cc @varkor

### CODE REVIEW

File: src/librustc_typeck/collect/type_of.rs [hunk 1]
```diff
@@ -235,38 +235,40 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
...
-                        // figure out which generic parameter it corresponds to and return
-                        // the relevant type.
-                        let generics = match path.res {
-                            Res::Def(DefKind::Ctor(..), def_id)
-                            | Res::Def(DefKind::AssocTy, def_id) => {
+                        // try to use the segment resolution if it is valid, otherwise we
+                        // default to the path resolution.
+                        let res = segment.res.filter(|&r| r != Res::Err).unwrap_or(path.res);
+                        let generics = match res {
+                            Res::Def(DefKind::Ctor(..) | DefKind::AssocTy, def_id) => {
```
Comment (eddyb) [inline src/librustc_typeck/collect/type_of.rs]: Is either of these needed anymore btw? Maybe just the `Ctor` case?

Comment (lcnr) [reply src/librustc_typeck/collect/type_of.rs]: Ctor seems to be required, removed assocty

File: src/librustc_typeck/collect/type_of.rs [hunk 2]
```diff
@@ -235,38 +235,40 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
...
                                 bug!("no arg matching AnonConst in path");
                             });
 
-                        // We've encountered an `AnonConst` in some path, so we need to
-                        // figure out which generic parameter it corresponds to and return
-                        // the relevant type.
-                        let generics = match path.res {
-                            Res::Def(DefKind::Ctor(..), def_id)
-                            | Res::Def(DefKind::AssocTy, def_id) => {
+                        // try to use the segment resolution if it is valid, otherwise we
```
Comment (eddyb) [inline src/librustc_typeck/collect/type_of.rs]: "try" -> "Try"

**Comment by lcnr** [comment]
ready for merge I guess.

**Pull request merged**